bandersnatch/fr/element.go

```diff
@@ -733,6 +734,28 @@ func (z *Element) SetBytesLE(e []byte) *Element {
 	bigIntPool.Put(vv)
 
 	return z
+
+}
+
+func (z *Element) SetBytesLECanonical(e []byte) (*Element, error) {
+	for i, j := 0, len(e)-1; i < j; i, j = i+1, j-1 {
+		e[i], e[j] = e[j], e[i]
+	}
+	// get a big int from our pool
+	vv := bigIntPool.Get().(*big.Int)
+	vv.SetBytes(e)
+
+	if vv.Cmp(&_modulus) != -1 {
+		return  nil, fmt.Errorf("not canonical")
+	}
+
+	// set big int
+	z.SetBigInt(vv)
+
+	// put temporary object back in pool
+	bigIntPool.Put(vv)
+
+	return z, nil
 }
```

Comment:
Here's a new API `SetBytesLECanonical`. It's a similar API to `SetBytesLE` but it forces the provided serialized integer passed as a parameter to be strictly less than the `_modulus`. 

This was discovered by the fuzzer creating examples of serialized `Multiproof` that didn't match the re-serialization of the struct. This is because the generated (and accepted) bytestream had `Fr` elements bigger than the modulus. And the serialization serialize them in "normalized" form. I'm not sure this would be a big deal, but probably we should only allow canonical serialization and deserialization.

Instead of creating this new API `SetBytesLECanonical`, I could also have added a new parameter `mustBeCanonical bool` to `SetBytesLE`, and avoid some duplication. The problem with that is that it would force `SetBytesLE` to return an error, which is unfortunate in other uses of `SetBytesLE` which can't never fail.

Other cases calling `SetBytesLE` are:
- Pulling 32 bytes from the FS transcript and generating a `Fr`.
- When mapping a Point to the base field, then transform the serialized Fp into Fr.

In both cases, we _want_ the wrapping behavior. And forcing these cases to handle an error that can only happen if that `mustBeCanonical` is set to `true` (which in these cases would be `false`) is unfortunate. If `false` the logic can't return an error, so we're bubbling a lot of unneeded errors in other parts of the code. Thus, I preferred to add the new `SetBytesLECanonical` which does return an error and clients should always handle that (and not ignore it).


Here's a fix to a discovered bug when fuzzing `Multiproof` serialization+deserialization.

The problem is similar to the one found yesterday. `SetBytesLE` was wrapping around the deserialized `e` in L735. The `SetBigInt(vv)` implementation (see L744) is doing a modulus operation if the integer is bigger than the modulus.

This creates the situation in which a serialized field element has potentially more than one allowed serialized bytestream. This potentially makes some of the `Multiproof` fields have a non-canonical way of being deserialized, which isn't ideal.

Now, we can't simply force everyone calling `SetBytesLE` to always provide a "normalized" representation since this is called in other places which actually _need_ this automatic "mod" behavior, some examples of those cases:
- We pull 32 random bytes from the FS transcript to generate a `Fr`. We clearly want this automatic modulus.
- In the `MapToBaseField` we map a `Point` to `Fp` and then the `Fp` to `Fr` which requires the wrapping behavior.

For that reason, I included the `mustBeCanonical` flag. if this is `true`, we check that the provided integer should be strictly less than the `_modulus`. Also recall that `vv.SetBytes` interprets bytes as an unsigned integer (so, it can't be negative).

The only thing that I don't like is touching this generated file; but I think it's already touched in multiple parts. If we ever regenerate this file, we'll notice this feature will be missing (or maybe impl in another gnark-crypto API?). At that point we should solve the issue (e.g: create a separate implementation to avoid touching the generated file; or plan for a PR in gnark-crypto and provide this feature natively).